File: Data Structures & Algorithms/binary-tree-diameter/submission-1.cpp
class Solution {
public:
    int diameter = 0;

    int height(TreeNode* node) {
        if (!node) {
            return 0;
        }

        int lHeight = height(node->left);
        int rHeight = height(node->right);

        // Update the global diameter with the maximum diameter found so far.
        // The diameter at the current node is the sum of the heights of its left and right subtrees.
        diameter = std::max(diameter, lHeight + rHeight);

        // Return the height of the current subtree.
        return 1 + std::max(lHeight, rHeight);
    }

    int diameterOfBinaryTree(TreeNode* root) {
        height(root);
        return diameter;
    }
};

File: Data Structures & Algorithms/max-water-container/submission-1.py
class Solution:
    def maxArea(self, heights: List[int]) -> int:
        l = 0
        r = len(heights) - 1
        length = r
        res = 0

        while l < r:
            area = min(heights[l], heights[r]) * length
            res = max(res, area)
            if heights[l] <= heights[r]:
                l += 1
                length -= 1
            else:
                r -= 1
                length -= 1
        return res